### DISCUSSION TITLE
Possible to avoid fully qualified column names in select statement?

### DISCUSSION THREAD

**Discussion by DanCardin** [sqlalchemy/sqlalchemy#9218] [Usage Questions] [answered]
I'm not entirely sure whether this will fully solve my problem; there is likely a long tail of unavoidable differences that I'll encounter.

tl;dr I'm trying to evaluate whether I use a SQLAlchemy query to create a view, and then exactly match the result of i.e. `select * from pg_views`.

Ignoring formatting, the main difference I'm seeing is that (which i guess this may be specific to postgres) it will omit the full tablename when rendering select statement columns, i.e. `select bar.baz from foo.bar`.

By contrast, SQLAlchemy will render the fully qualified column name i.e. `select foo.bar.baz from foo.bar`. Thus the question: is it possible to coerce SQLAlchemy to omit the fully qualified tablename, for the purposes of this comparison?

**Comment by zzzeek** [comment]
can you provide more specifics?  im not really understanding the question.   "it will omit the full tablename" , what is "it" ?  if you wrote the query "select * from view" in SQLAlchemy then there are no specific column names?

**Reply by DanCardin** [reply]
```
postgres@localhost:postgres> create schema wat
postgres@localhost:postgres> create table wat.foo (id integer);
postgres@localhost:postgres> create view meow as (select id FROM wat.foo)
postgres@localhost:postgres> select pg_get_viewdef(oid) from (select oid from pg_class where relname = 'meow') a
+------------------+
| pg_get_viewdef   |
|------------------|
|  SELECT foo.id   |
|    FROM wat.foo; |
+------------------+
```

whereas

```python
from sqlalchemy import table, column, select
from sqlalchemy.dialect import postgresql

foo = table('foo', column('id'), schema='wat')
print(str(select(foo).compile(dialect=postgresql.dialect())))
```

yields `SELECT wat.foo.id \nFROM wat.foo`. And here, ideally I would be able to convince sqlalchemy to omit `wat.` from `wat.foo.id`; or more generally, only qualify ambiguous column names, if possible.

...Ideally the database would have a better way to translate a query into a normalized form, like it's able to do for a view, but i'm not aware of a way to do that without actually creating the object.

**Reply by zzzeek** [reply]
> ideally I would be able to convince sqlalchemy to omit wat. from wat.foo.id

I'm not sure what the reasoning for this, if the SQL is valid.  is the SQL somehow not valid?

you can always apply an alias to the object, which will limit the extent to where the schema name is used. SQL Server sometimes needs this

    >>> print(str(select(foo.alias()).compile(dialect=postgresql.dialect())))
    SELECT foo_1.id  FROM wat.foo AS foo_1

if you don't want the phrase "wat." in there at all, like if "wat." were in your search_path, then just remove it from your `table()`.

**Reply by DanCardin** [reply]
The goal is to 100% match the output phrasing of `pg_get_viewdef`, which does not qualify schema names in the column names (presumably unless it's ambiguous).

The options, as I see them, are:

* Make postgres match sqlalchemy, but i haven't happened acrosss any way to do that
* Make sqlalchemy match postgres, which is the purpose of this inquery
* Coerce one or both into eachother after the fact, perhaps `sqlglot` could do this
* use alembic-utils' strategy of: **actually** create the view, in a transaction, just to obtain the canonical representation. This is fine...ish, if slow; but wouldn't work for anything which doesn't support transactional ddl.

**Reply by zzzeek** [reply]
> The goal is to 100% match the output phrasing of `pg_get_viewdef`, which does not qualify schema names in the column names (presumably unless it's ambiguous).

I guess you're trying to write some kind of alembic autogenerate thing, but having SQLAlchemy output match pg_get_viewdef exactly sounds fairly impractical.   Ways that would actually work include:

1. actually create a view in PG with your SELECT and then run it through that function.   Alembic uses this technique with PostgreSQL to compare server side default functions, I believe
2. use a tool such as [sqlparse](https://sqlparse.readthedocs.io/en/latest/) to get a syntax-tree for both statements and compare from that angle
3. **easiest** - use [alembic-utils](https://github.com/olirice/alembic_utils) which already features support for autogenerated PostgreSQL views.

**Comment by DanCardin** [comment, marked as answer]
Alright, I suppose i'll call this done then. there's not exactly a resolution beyond: no it's not possible. Actually creating the view is not cross-database compatible generally, because apparently numerous non-pg databases dont support transactional ddl; which is why alembic-utils only supports postgres.

**Reply by zzzeek** [reply]
it's not an easy problem that's for sure, though alembic includes for dialect-specific comparison schemes.   I'm not ready to have sqlparse as a dependency but it could be something that works as a third party thing.
